engine/dscache.go

```diff
@@ -0,0 +1,215 @@
+package engine
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	lru "github.com/golang/groupcache/lru"
+	ds "github.com/ipfs/go-datastore"
+	dsn "github.com/ipfs/go-datastore/namespace"
+	dsq "github.com/ipfs/go-datastore/query"
+)
+
+const linksCachePath = "/cache/links"
+
+// dsCache is a LRU-cache that stores keys in memory and values in a datastore.
+type dsCache struct {
+	dstore ds.Datastore
+	lru    *lru.Cache
+	lock   sync.RWMutex
+}
+
+// newDsCache creates a new dsCache instance.  The context is only used cancel
```

Comment:
i feel like we're written very similar caching wrappers around datastores before. i wonder if we can make a canonical one that gets reused.

Reply:
I know we have one around the blockstore that uses ARC cache or Two-Queue Cache.  I think cache eviction means something different to it though.  I will take a look to see if I can find others.